xbmc/guilib/GUITextLayout.cpp

```diff
@@ -256,56 +256,68 @@ void CGUITextLayout::UpdateStyled(const vecText &text, const std::vector<UTILS::
 }
 
 // BidiTransform is used to handle RTL text flipping in the string
-void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTRReadingOrder)
+void CGUITextLayout::BidiTransform(std::vector<CGUIString>& lines, bool forceLTRReadingOrder)
 {
-  for (unsigned int i=0; i<lines.size(); i++)
+  for (unsigned int i = 0; i < lines.size(); i++)
   {
-    CGUIString &line = lines[i];
+    CGUIString& line = lines[i];
+    unsigned int lineLength = line.m_text.size();
+    std::wstring logicalText;
+    vecText style;
 
-    // reserve enough space in the flipped text
-    vecText flippedText;
-    flippedText.reserve(line.m_text.size());
+    logicalText.reserve(lineLength);
+    style.reserve(lineLength);
 
-    character_t sectionStyle = 0xffff0000; // impossible to achieve
-    std::wstring sectionText;
+    // Separate the text and style for the input styled text
     for (const auto& it : line.m_text)
     {
-      character_t style = it & 0xffff0000;
-      if (style != sectionStyle)
+      logicalText.push_back((wchar_t)(it & 0xffff));
+      style.push_back(it & 0xffff0000);
+    }
+
+    // Allocate memory for logical to visual map and call bidi
+    int* logicalToVisualMap = new int[lineLength + 1];
+    std::wstring visualText = BidiFlip(logicalText, forceLTRReadingOrder, logicalToVisualMap);
+
+    vecText styledVisualText;
+    styledVisualText.reserve(lineLength);
+
+    // If memory allocation failed, fallback to text with no styling
+    if (!logicalToVisualMap)
+    {
+      for (unsigned int j = 0; j < visualText.size(); j++)
       {
-        if (!sectionText.empty())
-        { // style has changed, bidi flip text
-          std::wstring sectionFlipped = BidiFlip(sectionText, forceLTRReadingOrder);
-          for (unsigned int j = 0; j < sectionFlipped.size(); j++)
-            flippedText.push_back(sectionStyle | sectionFlipped[j]);
-        }
-        sectionStyle = style;
-        sectionText.clear();
+        styledVisualText.push_back(visualText[j]);
       }
-      sectionText.push_back((wchar_t)(it & 0xffff));
     }
-
-    // handle the last section
-    if (!sectionText.empty())
+    else
     {
-      std::wstring sectionFlipped = BidiFlip(sectionText, forceLTRReadingOrder);
-      for (unsigned int j = 0; j < sectionFlipped.size(); j++)
-        flippedText.push_back(sectionStyle | sectionFlipped[j]);
+      for (unsigned int j = 0; j < visualText.size(); j++)
+      {
+        styledVisualText.push_back(style[logicalToVisualMap[j]] | visualText[j]);
```

Comment:
I just realized that there is a mistake here. I used the value of logicalToVisualMap[j] as index of style[] which is in logical order.  While this works for most cases, it's not correct; it should be visualToLogicalMap[] instead.
```suggestion
        styledVisualText.push_back(style[visualToLogicalMap[j]] | visualText[j]);
```
To correct this I will rename logicalToVisualMap to visualToLogicalMap everywhere it's used and use the respective output from fribidi for it.